### PR TITLE
[LIEF-12747] Gzip body in Graphhopper requests

### PR DESCRIPTION
**Pull request by tiagotex**
```
Bandwidth reduction
If you create your own client, make sure it supports http/2 and gzipped responses for best speed.

If you use the Matrix or Route Optimization API and want to solve large problems, we recommend you to reduce bandwidth by compressing your POST request and specifying the header as follows: Content-Encoding: gzip.
```

https://docs.graphhopper.com/#section/Explore-our-APIs/API-Clients

### CODE REVIEW

File: lib/graphhopper/api_client.rb [hunk 1]
```diff
@@ -277,7 +279,8 @@ def build_request_body(header_params, form_params, body)
       else
         data = nil
```
Comment (adamniedzielski) [inline lib/graphhopper/api_client.rb]: What happens if we pass `nil` to `Zlib::Deflate.deflate`?

Comment (PragTob) [reply lib/graphhopper/api_client.rb]:
```
tobi@speedy:~$ irb
irb(main):001:0> require 'zlib'
=> true
irb(main):002:0> Zlib::Deflate.deflate nil
Traceback (most recent call last):
        5: from /home/tobi/.asdf/installs/ruby/2.6.0/bin/irb:23:in `<main>'
        4: from /home/tobi/.asdf/installs/ruby/2.6.0/bin/irb:23:in `load'
        3: from /home/tobi/.asdf/installs/ruby/2.6.0/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):2
        1: from (irb):2:in `deflate'
TypeError (no implicit conversion of nil into String)
```

Good catch Adam.

Comment (tiagotex) [reply lib/graphhopper/api_client.rb]: yeah, I didn't get that far, because gaphhopper keeps insisting that the request is not in GZIP format 😔

**Comment by adamniedzielski** [comment]
@tiagotex 5k shipments squad will take over the PR

File: lib/graphhopper/api/vrp_api.rb [hunk 2]
```diff
@@ -50,6 +50,8 @@ def post_vrp_with_http_info(key, body, opts = {})
       header_params['Accept'] = @api_client.select_header_accept(['application/json'])
       # HTTP header 'Content-Type'
       header_params['Content-Type'] = @api_client.select_header_content_type(['application/json'])
+      # HTTP header 'Content-Type'
```
Comment (adamniedzielski) [inline lib/graphhopper/api/vrp_api.rb]: ©️ ✂️ (copy-paste 😆 )

And do we event need these comments? They don't seem helpful.

File: lib/graphhopper/api_client.rb [hunk 3]
```diff
@@ -277,6 +278,11 @@ def build_request_body(header_params, form_params, body)
...
       end
+
+      if data && header_params['Content-Encoding'] == 'gzip'
```
Comment (adamniedzielski) [inline lib/graphhopper/api_client.rb]: I think that setting the header in one place and then reading it another to make a decision is a bit confusing and makes the code hard to follow.

What if we had a `use_compression` option for `GraphHopper:: ApiClient` that defaults to `false`? Then `GraphHopper:: VrpApi` can set it to true in `initialize`. I find the pattern of calling `select_header_content_type` a bit confusing, but we could follow it and implement `select_header_content_encoding` that returns the correct value based on `use_compression`. If we do this, all the compression logic is encapsulated in `GraphHopper:: ApiClient`.

What do you think?

File: lib/graphhopper/api/vrp_api.rb (hunk 2 above)
Comment (irmela) [reply lib/graphhopper/api/vrp_api.rb]: I don't think they are very useful but I find it easier to read the code when it is consistent.

File: lib/graphhopper/api_client.rb [hunk 4]
```diff
@@ -277,6 +281,11 @@ def build_request_body(header_params, form_params, body)
...
```
Comment (adamniedzielski) [inline lib/graphhopper/api_client.rb]: Hmm... I think that after introducing `use_compression` we can change `header_params['Content-Encoding'] == 'gzip'` to `use_compression`.

Comment (irmela) [reply lib/graphhopper/api_client.rb]: 🙈
